proposals/shmsize-tmpfs.md

```diff
@@ -0,0 +1,65 @@
+<!--
+Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License"). You may
+not use this file except in compliance with the License. A copy of the
+License is located at
+
+     http://aws.amazon.com/apache2.0/
+
+or in the "license" file accompanying this file. This file is distributed
+on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+express or implied. See the License for the specific language governing
+permissions and limitations under the License.
+-->
+
+# Support for ShmSize and Tmpfs Docker Parameters in ECS
+
+## Background
+
+Docker supports the parameters `--shm-size` and `--tmpfs` which sets the size of
+`/dev/shm` volume and creates an empty tmpfs mount within a container respectively.
+More information on these flags:
+* ShmSize: https://docs.docker.com/engine/reference/run/#runtime-constraints-on-resources
+* Tmpfs: https://docs.docker.com/engine/reference/run/#tmpfs-mount-tmpfs-filesystems
+
+## ECS Support for the flags
+
+### ShmSize
+A new field will be added for Docker's --shm-size flag under the linuxParameters
+field in ContainerDefinition of the Task definition. Through this field, customers
+will be able to set the value for `ShmSize` parameter in Docker's HostConfig that
+is passed during create container call.
+
+### Tmpfs
+A new field will be added for Docker's --tmpfs flag under the linuxParameters field
+in ContainerDefinition of the Task definition.  Through this field, customers will
+be able to set the value for `Tmpfs` parameter in Docker's HostConfig that is passed
+during the create container call.
+
+Tmpfs parameter takes in a container path as the mount point and other mount options
+for tmpfs including the mount size. Other options can be found in the man page for
+mount [here](http://man7.org/linux/man-pages/man8/mount.8.html).
+
+## Implications of using these fields during task placement decisions:
+
+Since both the flags consume memory in the instance in which their corresponding
+tasks are placed, their memory sizes will be accounted against the total memory
+available in the instance. As a result, the memory set for both the ShmSize and
+Tmpfs parameters will affect placement decisions for other tasks in the same instances.
+Size will be a mandatory parameter in the Tmpfs field since it is essential for
+making task placement decisions.
+
+CloudWatch Metrics would now include additional metrics on memory reserved for these
```

Comment:
Summarizing offline discussion - I think we'd want to introduce new reservation metrics only if we can have corresponding utilization metrics. If we're not going to track Shm and Tmpfs utilization independently, it might make sense to update existing metrics and not introduce new metrics.

Reply:
I don't think we want to create a new metric.  However, I do think we need to include shm and tmpfs memory in the existing reservation metric.  Customers use the reservation metric as a triggering metric in auto scaling policies; if we don't include shm and tmpfs memory, customers will be unable to use this metric as a way to accurately indicate the amount of available memory for future tasks in their clusters.

Reply:
`Tmpfs` and `/dev/shm` get used by processes as filesystem volumes that can be written to and not as their run-time memory and as such are treated differently by the kernel. For example, the OOM killer is not invoked if you try to create/extend a file beyond the tmfs volume size that was allocated as that's associated with the memory cgroup (specified by the `--memory` parameter). For these reasons, it doesn't make much sense to account for these in utilization metrics. However, as these will have an impact on task placement, we should definitely account for these in reservation metrics as pointed out by @samuelkarp's comment.

Reply:
Makes sense, thanks!